server/resp/src/test/java/org/infinispan/server/resp/RespDecoderTest.java

```diff
@@ -0,0 +1,87 @@
+package org.infinispan.server.resp;
+
+import static org.testng.AssertJUnit.assertEquals;
+import static org.testng.AssertJUnit.assertNotNull;
+
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.CompletionStage;
+
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.embedded.EmbeddedChannel;
+import io.netty.handler.codec.FixedLengthFrameDecoder;
+
+/**
+ * Test to ensure decoder works properly for various requests, types and replay
+ *
+ * @author William Burns
+ * @since 15.0
+ */
+@Test(groups = "functional", testName = "server.resp.RespDecoderTest")
+public class RespDecoderTest {
+
+   EmbeddedChannel channel;
+   Queue<Request> queuedCommands;
+
+   static class Request {
+      private final String command;
+      private final List<byte[]> arguments;
+
+      Request(String command, List<byte[]> arguments) {
+         this.command = command;
+         // Make a copy, because the decoder may reuse the list
+         this.arguments = new ArrayList<>(arguments);
+      }
+   }
+
+   @BeforeClass
+   public void beforeClass() {
+      queuedCommands = new ArrayDeque<>();
+      RespRequestHandler myRespRequestHandler = new RespRequestHandler() {
+         @Override
+         public CompletionStage<RespRequestHandler> handleRequest(ChannelHandlerContext ctx, String type, List<byte[]> arguments) {
+            queuedCommands.add(new Request(type, arguments));
+            return myStage;
+         }
+      };
+      channel = new EmbeddedChannel(new FixedLengthFrameDecoder(1), new RespDecoder(myRespRequestHandler));
+   }
+
+   @AfterClass
+   public void afterClass() {
+      channel.close();
+   }
+
+   @Test
+   public void testMixtureOfTypes() {
+      String commandName = "ALLTYPES";
+      ByteBuf buffer = Unpooled.copiedBuffer("*5\r\n+" + commandName + "\r\n$3\r\nkey\r\n+value\r\n:23\r\n$5\r\nworks\r\n", StandardCharsets.US_ASCII);
```

Comment:
Maybe including a negative number just for completeness.

Reply:
Sure